Chore/new tokens docs (#341)
momentohq/public-dev-docs: docs/develop/api-reference/auth-tokens.md

---
sidebar_position: 7
title: Authentication control token API reference information
sidebar_label: Authentication control tokens
description: Learn the authentication control token API calls you need to know about and how to use them with Momento services.
---

import { SdkExampleTabs } from "@site/src/components/SdkExampleTabs";
// This import is necessary even though it looks like it's un-used; The inject-example-code-snippet
// plugin will transform instances of SdkExampleTabs to SdkExampleTabsImpl
import { SdkExampleTabsImpl } from "@site/src/components/SdkExampleTabsImpl";

> # Authentication control API reference

Reviewer: Still conflicted about using `Authentication` here vs `Auth` to denote both AuthZ and AuthN

Author: Changed to auth.

<img src="/img/access-tokens.jpg" width="90%" alt="a technical illustration of a bank vault representing security, authorization, and authentication." />

> These APIs are used to manage Momento authentication tokens and access. Tokens can be scoped to have one or more roles to give access to one or more caches or topics.

Reviewer: I think `one or more permissions` is more accurate than `one or more roles`

Author: Correct. changed.

## GenerateAuthToken

Generates a new Momento authentication token with the specified permissions and expiry.

| Name            | Type                       | Description                                                                                       |
| --------------- |----------------------------|---------------------------------------------------------------------------------------------------|
| scope           | [TokenScope](#tokenscope)  | The permissions to grant to the new token. Pre-built TokenScope objects are provided by the SDKs. |

> | expiresIn       | Number \| ExpiresIn object | The number of seconds until the token expires or an ExpireIn object calling the `ExpireIn.never()` method.  |

Reviewer: It can also be a duration defined as minutes or hours by using ExpiresIn.minutes() or ExpiresIn.hours()

Author: added

| Name            | Type            | Description                                   |
| --------------- | --------------- | --------------------------------------------- |
| refreshToken    | String          | The refreshToken for the current auth token, acquired from the original call to `GenerateAuthToken`. |

<details>
  <summary>Method response object</summary>

* Success
  - `authToken`: string - the new auth token
  - `refreshToken`: string - a refresh token that can be used with the [RefreshAuthToken API](#refreshauthtoken) to refresh the token before it expires
  - `expiresAt`: Timestamp - the timestamp at which the token will expire
* Error

See [response objects](./response-objects) for specific information.

</details>

<SdkExampleTabs snippetId={'API_RefreshAuthToken'} />

<img src="/img/momento-auth-tokens.png" />

## TokenScope
| Name            | Type                                       | Description                                  |
| --------------- |--------------------------------------------| -------------------------------------------- |

> | permissions           | Array <[Permission](#permission-objects)\> | The permissions to grant to the new token.|

Reviewer: I think I used `Array` in my draft too but List might be a more generic term.

Author: Fixed.

A TokenScope is an array of [permission objects](#permission-objects). The array can have permissions that are of type [CachePermission](#cachepermission) or [TopicPermission](#topicpermission), and the array can contain [up to ten](../../manage/limits) permission objects. A permission only grants access to the Momento data plane APIs (e.g. `get`, `set`, etc.). When an auth token is created with multiple permission objects, any matching permission will grant access. For example, if a single token is created with two permission objects:

1. A permission object that allows ReadWrite access to all caches for the account 
2. A permission object that allows ReadOnly access to cache `foo`

In this case, the token will still allow use of data manipulation APIs (e.g. `set`, `delete`, `DictionarySetFields`, etc.) on cache `foo` because of the first permission.

## Permission Objects

> These objects define the specific role with cache or topic information and then assigned to a [TokenScope](#tokenscope). 

Reviewer: missing `are` in `then assigned`?

### CachePermission
A component of a [TokenScope](#tokenscope) object that defines permissions for a cache.

| Name            | Type                  | Description                                                                           |
| --------------- |-----------------------|---------------------------------------------------------------------------------------|
| role           | ReadOnly \| ReadWrite | The type of access granted by the permission.                                         |

> | cache          | `AllCaches` \| String | A permission can be restricted to a select cache by name or AllCaches built-in value. |

Reviewer: See my other comment about `CacheName` / `TopicName` interfaces

For role, using `CacheRole.ReadOnly` allows access to all read data plane APIs on caches (e.g. `get`, `DictionaryGetField`, etc.) defined in the CacheSelector. Using `CacheRole.ReadWrite` allows access for all read and write data plane APIs on the caches defined in the CacheSelector. Custom roles are not supported.

For cache, the value can be the built-in `AllCaches` or a string value containing the name of the cache this permission is for.

### TopicPermission
A component of a [TokenScope](#tokenscope) object that defines permissions for a token.

| Name            | Type                              | Description                                                                          |
| --------------- |-----------------------------------|--------------------------------------------------------------------------------------|
| role           | SubscribeOnly \| PublishSubscribe | The type of access granted by the permission.               |

> | cache          | `AllCache` \| String              | A permission can be restricted to a select cache by name or AllCaches built-in value.              |

Reviewer: It can also be an object that satisfies the CacheName interface ` { name: string }`. The string is a shorthand. We may add other types of selectors like Prefix in future.
Similar comment for `topic` which can be a TopicName

Author: Not saying no, but for simplicity's sake, why would anyone ever use the object when they could just use a simple string? My thought is if in the future we change things, then we could put that in. Thoughts?

Reviewer: Because I am not sure we can expose this as a string in all SDKs - we may end up having CacheName/TopicName classes in other languages.

Reviewer: This documentation is common to all SDKs so it should define the fields logically rather than use specific data types.

Author: That makes far more sense to me and was not clear before. will change!

| topic          | `AllCache` \| String              | A permission can be restricted to a select topic by name or AllTopics built-in value. |

For role, there are two managed roles to assign, `TopicRole.PublishSubscribe` and `TopicRole.SubscribeOnly`. Custom roles are not supported. Using the SubscribeOnly role allows only subscribing to topics, whereas using the PublishSubscribe role allows publishing and subscribing to topics.

> For cache, only topics within that cache's namespace are allowed by the permission. This can be set to the built-in `AllCache` value or a string specifically naming a cache.

Reviewer: typo - should be `AllCaches`

> For topic, this can be set to the built-in `AllTopics` value, which gives access to all topics in the cache(s) defined in cache or it can be a string with a specific topic name.

Reviewer: Should cache be highlighted with backticks in `defined in cache`?

#### TopicScope examples
This is an example of creating a TokenScope with just CachePermissions.

> ```javascript

Reviewer: We shouldn't have code directly embedded here - I think you already pointed out it is temporary but just reiterating so we don't forget.

Reviewer: docs code snippets PR: https://github.com/momentohq/client-sdk-javascript/pull/623

Author: Yep. There is a plan to get this using the SDK repo examples, but we need to kick this out the door.

const CachePermissions = {
    permissions: [
        {
            role: CacheRole.ReadWrite, // Managed role
            cache: "MyCache" // grants access to a specific cache
        },
        {
            role: CacheRole.ReadOnly, // Managed role
            cache: AllCaches // Built-in value for access to all caches in the account.
        },
    ],
};
```

This is an example of creating a TokenScope with just TopicPermissions.

```javascript
const TopicsPermissions = {
    permissions: [
      {
        role: TopicRole.PublishSubscribe, // Managed role
        cache: 'the-great-wall', // grants access to a specific cache
        topic: 'highlights', // grants access to a specific topic
      },
      {
        role: TopicRole.SubscribeOnly, // Built-in role
        cache: AllCaches, // This is a built-in value for access to all caches in the account
        topic: AllTopics, // This is a built-in value for access to all topic in the listed cache(s).
      },
    ],
};
```

:::tip

If you have any questions not answered here, please jump on [our Discord server](https://discord.gg/2c24SK6W) and ask our experts in the support channel.

:::

## FAQ

> <details>

Reviewer 2: Another question worthwhile adding is that we do now allow tokens with write only or publish only permissions

Author: I will add this.

Reviewer 2: Thank you

---
sidebar_position: 7

> title: Auth control token API reference information

Reviewer: ```suggestion
title: Access control token API reference information
```

Author: I thought you wanted things to say Auth.

> title: Access control token API reference information

Reviewer: @danielamiao - are you ok with using Access Control Tokens as the main term? cc @rishtigupta and @bruuuuuuuce for the console wording

Reviewer 3: Fine Grained Access Control (FGAC) is better

Author: Discussed offline, resolving.

> sidebar_label: Auth control tokens

Reviewer: ```suggestion
sidebar_label: Access control tokens
```

> description: Learn the auth control token API calls you need to know about and how to use them with Momento services.

Reviewer: ```suggestion
description: Learn the access control token API calls you need to know about and how to use them with Momento services.
```

---

import { SdkExampleTabs } from "@site/src/components/SdkExampleTabs";
// This import is necessary even though it looks like it's un-used; The inject-example-code-snippet
// plugin will transform instances of SdkExampleTabs to SdkExampleTabsImpl
import { SdkExampleTabsImpl } from "@site/src/components/SdkExampleTabsImpl";

> # Auth control API reference

Reviewer: ```suggestion
# Access control token API reference
```

<img src="/img/access-tokens.jpg" width="90%" alt="a technical illustration of a bank vault representing security, authorization, and authentication." />

> These APIs are used to manage Momento auth tokens and access. Tokens can be scoped to have one or more permissions to give access to one or more caches or topics.

Reviewer: ```suggestion
These APIs are used to manage Momento access control tokens and access. Tokens can be scoped to have one or more permissions to give access to one or more caches or topics.
```

Author: ok, this one is a problem. most everywhere in the docs we call these tokens "Momento auth tokens", not Momento access control tokens. Has this changed and we need to change this everywhere?

Reviewer: We cannot get rid of the `auth token` terminology. It is everywhere including api names. I suggest we add a transition phrase like `Momento authentication and authorization token for fine grained access control, henceforth referred to as auth token` at the beginning of the page.

## GenerateAuthToken

> Generates a new Momento auth token with the specified permissions and expiry.

Reviewer: ```suggestion
Generates a new Momento authentication and authorization token for access control with the specified permissions and expiry, henceforth referred to as auth token.
```

| Name            | Type                       | Description                                                                                                                                               |
| --------------- |----------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------|
| scope           | [TokenScope](#tokenscope)  | The permissions to grant to the new token. Pre-built TokenScope objects are provided by the SDKs.                                                         |

> | expiresIn       | Number \| ExpiresIn object | The number of seconds until the token expires or an ExpireIn object calling the `ExpireIn.never()`, `ExpireIn.minutes()`, or `ExpireIn.hours()` methods. |

Reviewer: ```suggestion
| expiresIn       | Number \| ExpiresIn object | The number of seconds until the token expires or an ExpiresIn object representing a duration by calling the `ExpiresIn.never()`, `ExpiresIn.seconds()`, `ExpiresIn.minutes()`, or `ExpiresIn.hours()` methods. |
```

| --------------- |--------------------------------------------| -------------------------------------------- |
| permissions           | Array <[Permission](#permission-objects)\> | The permissions to grant to the new token.|

A TokenScope is an array of [permission objects](#permission-objects). The array can have permissions that are of type [CachePermission](#cachepermission) or [TopicPermission](#topicpermission), and the array can contain [up to ten](../../manage/limits) permission objects. A permission only grants access to the Momento data plane APIs (e.g. `get`, `set`, etc.). When an auth token is created with multiple permission objects, any matching permission will grant access. For example, if a single token is created with two permission objects:

1. A permission object that allows ReadWrite access to all caches for the account 
2. A permission object that allows ReadOnly access to cache `foo`

In this case, the token will still allow use of data manipulation APIs (e.g. `set`, `delete`, `DictionarySetFields`, etc.) on cache `foo` because of the first permission.

## Permission Objects

> These objects define the specific role with cache or topic information and then assigned to a [TokenScope](#tokenscope). 

Reviewer: ```suggestion
These objects define the specific role with cache or topic information and are assigned to a [TokenScope](#tokenscope). 
```
